// nexus/dist/runtime/schema/schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.create = void 0;
const tslib_1 = require("tslib");
const NexusSchema = tslib_1.__importStar(require("@nexus/schema"));
const chalk_1 = tslib_1.__importDefault(require("chalk"));
const nexus_schema_stateful_1 = require("../../lib/nexus-schema-stateful");
const Process = tslib_1.__importStar(require("../../lib/process"));
const Scalars = tslib_1.__importStar(require("../../lib/scalars"));
const DevMode = tslib_1.__importStar(require("../dev-mode"));
const utils_1 = require("../utils");
const logger_1 = require("./logger");
const settings_1 = require("./settings");
const settings_mapper_1 = require("./settings-mapper");
function createLazyState() {
    return {
        contextContributors: [],
        plugins: [],
        scalars: {},
    };
}
function create(appState) {
    appState.schemaComponent = createLazyState();
    const statefulNexusSchema = nexus_schema_stateful_1.createNexusSchemaStateful();
    const settings = settings_1.createSchemaSettingsManager();
    const api = Object.assign(Object.assign({}, statefulNexusSchema.builders), { use(plugin) {
            utils_1.assertAppNotAssembled(appState, 'app.schema.use', 'The Nexus Schema plugin you used will be ignored.');
            appState.schemaComponent.plugins.push(plugin);
        },
        addToContext(contextContributor) {
            appState.schemaComponent.contextContributors.push(contextContributor);
        },
        middleware(fn) {
            api.use(NexusSchema.plugin({
                // TODO: Do we need to expose the name property?
                name: 'local-middleware',
                onCreateFieldResolver(config) {
                    return fn(config);
                },
            }));
        } });
    return {
        public: api,
        private: {
            settings: settings,
            reset() {
                statefulNexusSchema.state.types = [];
                statefulNexusSchema.state.scalars = {};
                appState.schemaComponent.contextContributors = [];
                appState.schemaComponent.plugins = [];
                appState.schemaComponent.scalars = {};
            },
            beforeAssembly() {
                appState.schemaComponent.scalars = statefulNexusSchema.state.scalars;
            },
            assemble(plugins) {
                const nexusSchemaConfig = settings_mapper_1.mapSettingsAndPluginsToNexusSchemaConfig(plugins, settings.data);
                nexusSchemaConfig.types.push(...statefulNexusSchema.state.types);
                nexusSchemaConfig.plugins.push(...appState.schemaComponent.plugins);
                const { schema, missingTypes } = NexusSchema.core.makeSchemaInternal(nexusSchemaConfig);
                return { schema, missingTypes };
            },
            checks() {
                assertNoMissingTypesDev(appState.assembled.schema, appState.assembled.missingTypes);
                // TODO: We should separate types added by the framework and the ones added by users
                if (statefulNexusSchema.state.types.length === 2 &&
                    statefulNexusSchema.state.types.every((t) => Scalars.builtinScalars[t.name] !== undefined)) {
                    logger_1.log.warn(emptyExceptionMessage());
                }
            },
        },
    };
}
exports.create = create;
function emptyExceptionMessage() {
    return `Your GraphQL schema is empty. This is normal if you have not defined any GraphQL types yet. If you did however, check that your files are contained in the same directory specified in the \`rootDir\` property of your tsconfig.json file.`;
}
function assertNoMissingTypesDev(schema, missingTypes) {
    const missingTypesNames = Object.keys(missingTypes);
    if (missingTypesNames.length === 0) {
        return;
    }
    const schemaTypeMap = schema.getTypeMap();
    const schemaTypeNames = Object.keys(schemaTypeMap).filter((typeName) => !NexusSchema.core.isUnknownType(schemaTypeMap[typeName]));
    if (DevMode.isDevMode()) {
        missingTypesNames.map((typeName) => {
            const suggestions = NexusSchema.core.suggestionList(typeName, schemaTypeNames);
            let suggestionsString = '';
            if (suggestions.length > 0) {
                suggestionsString = ` Did you mean ${suggestions
                    .map((s) => `"${chalk_1.default.greenBright(s)}"`)
                    .join(', ')} ?`;
            }
            logger_1.log.error(`Missing type "${chalk_1.default.redBright(typeName)}" in your GraphQL Schema.${suggestionsString}`);
        });
    }
    else {
        Process.fatal(`Missing types ${missingTypesNames.map((t) => `"${t}"`).join(', ')} in your GraphQL Schema.`, { missingTypesNames });
    }
}
